pkg/broker/broker_test.go

```diff
@@ -64,3 +65,176 @@ func ExampleBrokerService_IsEnabled() {
+	// Output: "Builtin!" <nil>
+	// "Custom!" <nil>
+	// Plan ID "missing-plan" could not be found
+}
+
+func TestBrokerService_UserDefinedPlans(t *testing.T) {
+	cases := map[string]struct {
+		Value       interface{}
+		PlanCount   int
+		ExpectError bool
+	}{
+		"default-no-plans": {
+			Value:       nil,
+			PlanCount:   0,
+			ExpectError: false,
+		},
+		"single-plan": {
+			Value:       `[{"id":"aaa"}]`,
+			PlanCount:   1,
+			ExpectError: false,
+		},
+		"bad-json": {
+			Value:       `42`,
+			PlanCount:   0,
+			ExpectError: true,
+		},
+		"multiple-plans": {
+			Value:       `[{"id":"aaa"},{"id":"bbb"}]`,
+			PlanCount:   2,
+			ExpectError: false,
+		},
+	}
+
+	service := BrokerService{
+		Name: "left-handed-smoke-sifter",
+		DefaultServiceDefinition: `{"id":"abcd-efgh-ijkl"}`,
+	}
+
+	for tn, tc := range cases {
+		viper.Set(service.UserDefinedPlansProperty(), tc.Value)
+		plans, err := service.UserDefinedPlans()
+
+		// Check errors
+		hasErr := err != nil
+		if hasErr != tc.ExpectError {
+			t.Errorf("%s) Expected Error? %v, got error: %v", tn, tc.ExpectError, err)
+		}
+
+		// Check IDs
+		if len(plans) != tc.PlanCount {
+			t.Errorf("%s) Expected %d plans, but got %d (%v)", tn, tc.PlanCount, len(plans), plans)
+		}
+
+		// Reset Environment
+		viper.Set(service.UserDefinedPlansProperty(), nil)
+	}
+}
+
+func TestBrokerService_CatalogEntry(t *testing.T) {
+	cases := map[string]struct {
+		UserDefinition interface{}
+		UserPlans      interface{}
+		PlanCount      int
+		ExpectError    bool
+	}{
+		"no-customization": {
+			UserDefinition: nil,
+			UserPlans:      nil,
+			PlanCount:      0,
+			ExpectError:    false,
+		},
+		"custom-definition": {
+			UserDefinition: `{"id":"abcd-efgh-ijkl", "plans":[{"id":"zzz"}]}`,
+			UserPlans:      nil,
+			PlanCount:      1,
+			ExpectError:    false,
+		},
+		"custom-plans": {
+			UserDefinition: nil,
+			UserPlans:      `[{"id":"aaa"},{"id":"bbb"}]`,
+			PlanCount:      2,
+			ExpectError:    false,
+		},
+		"custom-plans-and-definition": {
+			UserDefinition: `{"id":"abcd-efgh-ijkl", "plans":[{"id":"zzz"}]}`,
+			UserPlans:      `[{"id":"aaa"},{"id":"bbb"}]`,
+			PlanCount:      3,
+			ExpectError:    false,
+		},
+		"bad-definition-json": {
+			UserDefinition: `333`,
+			UserPlans:      nil,
+			PlanCount:      0,
+			ExpectError:    true,
+		},
+		"bad-plan-json": {
+			UserDefinition: nil,
+			UserPlans:      `333`,
+			PlanCount:      0,
+			ExpectError:    true,
+		},
+	}
+
+	service := BrokerService{
+		Name: "left-handed-smoke-sifter",
+		DefaultServiceDefinition: `{"id":"abcd-efgh-ijkl"}`,
+	}
+
+	for tn, tc := range cases {
+		viper.Set(service.DefinitionProperty(), tc.UserDefinition)
+		viper.Set(service.UserDefinedPlansProperty(), tc.UserPlans)
+
+		srvc, err := service.CatalogEntry()
+		hasErr := err != nil
+		if hasErr != tc.ExpectError {
+			t.Errorf("%s) Expected Error? %v, got error: %v", tn, tc.ExpectError, err)
+		}
+
+		if err == nil && len(srvc.Plans) != tc.PlanCount {
```

Comment:
Same comment as above, but additionally verifying correct propagation of the user defined service ID.